_posts/2018-07-10-Play Framework-in-Scala–Action-Composition.md

```diff
@@ -0,0 +1,256 @@
+
+```scala
+def invokeBlock[A](request: Request[A],
+                   block: Request[A] => Future[Result]): Future[Result] = {
+  Logger.debug(s"Just received the following request: $request")
+  block(request)
+}
+```
+
+Often it is convenient that a custom ActionBuilder, via its invokeBlock method, 
+takes a request, does something with it, builds an instance of another type, and <strong>only then</strong> applies the block function to that new instance.
+Authentication of requests is a good example. Here, a particular request, if authenticated, is normally associated with a User. It is therefore useful to obtain the corresponding instance of User and pass it along, meaning, applying the block function to that instance.
+This is useful because now the Actions that are build through this ActionBuilder may count on the existence of an instance of User instead of a raw request. 
+This of course changes the signature of the block function that must be provided for the posterior creation of each Action.
+A naive AuthenticationActionBuilder could be:
+
+```scala
+class AuthenticatedAction[B](val parser: BodyParser[B])
+                            (implicit val executionContext: ExecutionContext) 
+                            extends ActionBuilder[AuthenticatedUser, B] {
+
+override def invokeBlock[A](request: Request[A],
+                            block: AuthenticatedUser[A] => Future[Result]): Future[Result] = {
+ val userId: Option[Int] = request.headers.get("user_id")
+ val password: Option[String] = request.headers.get("password")
+
+ val user: Option[User] = DataBase.validate(userId, password)
+
+ user.fold {
+   Future.successful(BadRequest("Authentication failed"))
+ }{ user =>
+   val authenticatedUser = AuthenticatedUser(user, request)
+   block(authenticatedUser)
+ }
+ }
+}
+```
+
+Notice above, that the type parameter of the ActionBuilder changed to  <code>AuthenticatedUser</code> and correspondingly the signature of the block function on the invokeBlock method is <small><code> block: AuthenticatedUser[A] => Future[Result] </code></small>. 
+It is up to us to define type AuthenticatedUser[_]. In this case, it could something along:
+```scala
+case class AuthenticatedUser[A](user: User, request: Request[A]) extends WrappedRequest(request)
+```
+
+Where type <code>User</code> is our internal representation of a user.
+AuthenticatedUser[_] must allow a type constructor, as demanded by the signature of the invokeBlock method of the trait ActionBuilder. 
+Also, as mentioned in the references, and explained further below, it is advantageous that this type parameter R[_] extends WrappedRequest[_]; which is a type provided by Play.
+
+As a side observation, you can do a trick not to need a type constructor:
+
+```scala
+case class AuthenticatedUserHelper(user: User)
+type AuthenticatedUser[_] = AuthenticatedUserHelper
+```
```

Comment:
Why would you need this helper?

Reply:
When you don't want a type constructor but rather a normal type.
This was something I remembered. 

The signature of the `ActionBuilder` requires a type constructor.  Line 150 above allows you to bypass this constraint. The type constructor required by Play is,  I believe, to abstract over the body type of the request.
This is then useful when you don't want this abstraction. When you don't want to use the body of the request on the downstream actions created by the ActionBuilder. 
Notice that the type constructor is only useful when you have a `Request[A]`as in:
```
case class AuthenticatedUser[A](user: User, request: Request[A]) extends WrappedRequest(request)
```

Not sure how useful this would be in practice.